Code cleanup

NewbieOrange opened the pull request:
This PR removes redundant modifiers and initializations, unused variables, incorrect javadoc references, etc.

The `static` modifiers in inner interfaces are removed since they are static by default, and removing them should not break binary compatibility (the md5sums are the same). Let me know if those are preferred to be kept.

NewbieOrange commented:
Also there are a lot of `@SuppressWarnings("deprecation")` removed since we're at language level 1.5 and should not trigger any warnings.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -7732,7 +7724,6 @@ void initFromMixin(UsageMessageSpec mixin, CommandSpec commandSpec) {
                 if (initializable(sortOptions, mixin.sortOptions(), DEFAULT_SORT_OPTIONS))                             {sortOptions = mixin.sortOptions();}
                 if (initializable(synopsisHeading, mixin.synopsisHeading(), DEFAULT_SYNOPSIS_HEADING))                 {synopsisHeading = mixin.synopsisHeading();}
                 if (initializable(synopsisSubcommandLabel, mixin.synopsisSubcommandLabel(), DEFAULT_SYNOPSIS_SUBCOMMANDS)) {synopsisHeading = mixin.synopsisHeading();}
-                if (initializable(width, mixin.width(), DEFAULT_USAGE_WIDTH))                                          {width = mixin.width();}
```
Why is this removed?

NewbieOrange replied:
This is never reached. initializable replaced the null varaible with default value, but an int is never null.

NewbieOrange replied:
If we want to keep this, we will need to introduce `initializable`s with each primitive type.

remkop replied:
Please revert this one; the warning that this condition is always `false` should go away with the latest master.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -5138,9 +5133,8 @@ public Help create(CommandSpec commandSpec, Help.ColorScheme colorScheme) {
             try {
-                @SuppressWarnings("deprecation") // Class.newInstance is deprecated in Java 9
-                T result = cls.newInstance();
-                return result;
```
This `@SuppressWarnings("deprecation")` annotation suppresses the warning one gets when building the project with Java 9. I think it is better to keep it.

remkop commented on src/main/java/picocli/AutoComplete.java:
```diff
@@ -235,7 +235,7 @@ public void run() {
         }
     }
 
-    private static interface Function<T, V> {
```
No problem, I am okay with removing `static` for inner interfaces.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -1470,7 +1470,7 @@ public void clearExecutionResults() {
      * @deprecated Use {@link IExecutionStrategy} instead.
      * @since 2.0 */
-    @Deprecated public static interface IParseResultHandler {
+    @Deprecated public interface IParseResultHandler {
```
Again, no problem removing `static` for inner interfaces.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -1758,7 +1758,7 @@ protected R throwOrExit(ExecutionException ex) {
      * @since 2.0 */
     @Deprecated public static class DefaultExceptionHandler<R> extends AbstractHandler<R, DefaultExceptionHandler<R>> implements IExceptionHandler, IExceptionHandler2<R> {
         public List<Object> handleException(ParameterException ex, PrintStream out, Help.Ansi ansi, String... args) {
-            internalHandleParseException(ex, new PrintWriter(out, true), Help.defaultColorScheme(ansi)); return Collections.<Object>emptyList(); }
```
I think I introduced this to get rid of a "raw collections" warning when building with Java 8 or higher... Probably better to keep the generified form with `<Object>`.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -2036,7 +2035,6 @@ public int execute(String... args) {
         } catch (ExecutionException ex) {
             try {
-                @SuppressWarnings("unchecked")
                 Exception cause = ex.getCause() instanceof Exception ? (Exception) ex.getCause() : ex;
```
`Exception.getCause` returns a `Throwable`. We never catch Throwable so we can safely downcast it to Exception, but I believe we do need to keep this annotation to suppress the compiler warning.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -2140,7 +2138,7 @@ private int resolveExitCode(int exitCodeOnSuccess, R executionResult, List<IExit
                 try {
                     int exitCode = generator.getExitCode();
-                    if ((exitCode > 0 && exitCode > result) || (exitCode < 0 && result <= 0 && exitCode < result)) {
+                    if ((exitCode > 0 && exitCode > result) || (exitCode < result && result <= 0)) {
```
Nice simplification!

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -2217,7 +2214,7 @@ private int resolveExitCode(int exitCodeOnSuccess, R executionResult, List<IExit
         protected List<IExitCodeGenerator> extractExitCodeGenerators(ParseResult parseResult) {
             if (parseResult.commandSpec().userObject() instanceof IExitCodeGenerator) {
-                return Arrays.asList((IExitCodeGenerator) parseResult.commandSpec().userObject());
+                return Collections.singletonList((IExitCodeGenerator) parseResult.commandSpec().userObject());
```
Nice. I overlooked that this API exists.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -8431,7 +8422,7 @@ private String scopeString() {
                 StringBuilder current = splittable;
                 boolean escaping = false, inQuote = false;
-                for (int ch = 0, i = 0; i < value.length(); i += Character.charCount(ch)) {
+                for (int ch, i = 0; i < value.length(); i += Character.charCount(ch)) {
```
personally I prefer to explicitly initialize all variables, but this is a minor quibble.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -8473,7 +8464,7 @@ private String scopeString() {
                 StringBuilder result = new StringBuilder();
                 boolean escaping = false, inQuote = false, skip = false;
-                for (int ch = 0, i = 0; i < part.length(); i += Character.charCount(ch)) {
+                for (int ch, i = 0; i < part.length(); i += Character.charCount(ch)) {
```
Same as above.

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -10038,7 +10025,6 @@ public MethodParam(Method method, int paramIndex) {
             }
             @Override public Annotation[] getDeclaredAnnotations() { return method.getParameterAnnotations()[paramIndex]; }
             @Override public void setAccessible(boolean flag) throws SecurityException { method.setAccessible(flag); }
-            @SuppressWarnings("deprecation")
```
`Method.isAccessible` will give deprecation warnings when building on Java 9+.
I believe we want to suppress this.
Try this:

```
%JAVA_11_HOME%\bin\javac -Xlint:all -d c:\Temp picocli\CommandLine.java
```

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -14943,7 +14921,7 @@ public IParamLabelRenderer createDefaultParamLabelRenderer() {
         static class DefaultOptionRenderer implements IOptionRenderer {
-            private String requiredMarker = " ";
+            private String requiredMarker;
             private boolean showDefaultValues;
```
Can we make `requiredMarker` and `showDefaultValues` `final` fields?

remkop commented on src/main/java/picocli/CommandLine.java:
```diff
@@ -15063,7 +15041,7 @@ private Text createLongOptionText(OptionSpec option, IParamLabelRenderer rendere
         static class DefaultParameterRenderer implements IParameterRenderer {
-            private String requiredMarker = " ";
+            private String requiredMarker;
             private boolean showDefaultValues;
```
Can we make `requiredMarker` and `showDefaultValues` `final` fields?

NewbieOrange replied:
Yes, I added a commit with some other private fields being made final.

NewbieOrange commented:
I have checked the other removed `@SuppressWarnings` does not introduce any new warning when compiled with OpenJDK 11.0.7 and -Xlint:all.

NewbieOrange commented:
The added final modifiers seem to break the codegen graalvm-aot tests, since the generated fields have an entra `"allowWrite" : true`.

NewbieOrange commented:
On a side note, I think we should also remove the broken kloudtek logo (HTTPS cert expired) and machine-translated article (从Commons CLI迁移到Picocli) from README.

remkop commented:
> I have checked the other removed @SuppressWarnings does not introduce any new warning when compiled with OpenJDK 11.0.7 and -Xlint:all.

Thank you!

> The added final modifiers seem to break the codegen graalvm-aot tests, since the generated fields have an entra `"allowWrite" : true`.
> 
> Edit: The field should not be made final, reverted.

Interesting! I’ll take a look tomorrow. 

> On a side note, I think we should also remove the broken kloudtek logo (HTTPS cert expired) and machine-translated article (从Commons CLI迁移到Picocli) from README.

Is the translation that bad? :-)
Even if it is, does it reflect badly on picocli? 
I was thinking it’s better to have more links than fewer, to spread awareness of the library.

NewbieOrange commented:
> > The added final modifiers seem to break the codegen graalvm-aot tests, since the generated fields have an entra `"allowWrite" : true`.
> > Edit: The field should not be made final, reverted.
> 
> Interesting! I’ll take a look tomorrow.
> 
This is due to a private field being made final, but it is accessed by reflection from annotations. I have removed the final modifier that caused the issue.

> > On a side note, I think we should also remove the broken kloudtek logo (HTTPS cert expired) and machine-translated article (从Commons CLI迁移到Picocli) from README.
> 
> Is the translation that bad? :-)
> Even if it is, does it reflect badly on picocli?
> I was thinking it’s better to have more links than fewer, to spread awareness of the library.

The translation is simply google translate I think. It is not unreadable tho, but some code snippets are being 'translated', and it is also not the original post. The original post is here (the content is the same): https://blog.csdn.net/Tybyqi/article/details/85787550

NewbieOrange commented:
>Can these be final too?

Made the fields final.

> When I tried compiling on java 11 with -Xlint:all, I saw deprecation warnings for method.isAccessible.
You did not get this?

No.. probably due to difference in build options (different distro, etc), I have added the @SupressWarnings annotation anyway.

> Please revert this one; the warning that this condition is always false should go away with the latest master.

Done.

remkop commented:
Merged into master.
Thank you for the contribution! Much appreciated!
